### ISSUE TITLE
[BUG]: integration crashing on 2022.11.1

### ISSUE THREAD

**Issue by robertklep** [plugwise/plugwise-beta#313]
### Describe the bug.

The integration is crashing during HA startup. I was initially using the HA integration but installed the beta to test, and it's crashing with the same error:

```
2022-11-04 12:04:13.762 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Anna for plugwise
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 365, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/plugwise/__init__.py", line 16, in async_setup_entry
    return await async_setup_entry_gw(hass, entry)
  File "/config/custom_components/plugwise/gateway.py", line 76, in async_setup_entry_gw
    api.get_all_devices()
  File "/usr/local/lib/python3.10/site-packages/plugwise/smile.py", line 160, in get_all_devices
    self._all_device_data()
  File "/usr/local/lib/python3.10/site-packages/plugwise/smile.py", line 114, in _all_device_data
    self.update_for_cooling(self.gw_devices[device_id])
  File "/usr/local/lib/python3.10/site-packages/plugwise/smile.py", line 80, in update_for_cooling
    max_setpoint = self._sched_setpoints[1]
IndexError: list index out of range
```

This started with 2022.11, I was running 2022.10 before and it worked fine.

FWIW, I'm running the Elga firmware (2018-02-08T11:15:53+01:00), Elga is not set up for cooling.

### What version of Home Assistant Core has the issue?

2022.11.1

### What was the last working version of Home Assistant Core?

2022.10

### What type of installation are you running?

Home Assistant Container

### How did you install plugwise-beta?

Cloned from GitHub

### What kind of Plugwise device is having issues?

Smile: Anna

### What firmware version is your Plugwise product at?

4.3.12

### Logging

It doesn't seem to even get there, the component is crashing before that.

### Diagnostics information

Can't seem to find this?

**Comment by bouwew** [comment]
Are you using the HA Core Plugwise integration that's in Core version 2022.11.0/1? Or are using Plugwise-beta?

**Comment by robertklep** [comment]
@bouwew the stack trace I posted is from the beta (but the same issue occurs with the HA Core integration).

**Comment by bouwew** [comment]
Ok, you have an Anna connected to an Elga, and the Elga is not set up for cooling. What does this mean exactly? You do have the cooling-feature installed?

**Comment by bouwew** [comment]
I'm looking at our existing userdata, you have the Elga-support feature, the cooling-feature is not for the Elga.
And our code concludes: elga-support present = cooling present. And in your case that's not correct.
The key question for me then is: how do I determine correctly with the Elga that the cooling-function is available, or not.

**Comment by bouwew** [comment]
Can you please capture the xml-data present at http://<smile_ip>/core/domain_objects?
Please do this via right-click -> Save as -> .xml

Please send the resulting file to my email: bouwe at westerdijk point info

**Comment by robertklep** [comment]
@bouwew with my previous thermostat (the Honeywell) I set the Elga to "alleen verwarmen" since the setup of my central heating doesn't allow for cooling. The reason I mention cooling is because of the error: ```
  File "/usr/local/lib/python3.10/site-packages/plugwise/smile.py", line 80, in update_for_cooling
    max_setpoint = self._sched_setpoints[1]
```

I'll e-mail you the Smile data ASAP.

**Comment by bouwew** [comment]
Xml-data has arrived.

How do you set the Elga to "alleen verwarmen"? Via a toggle switch?
I do see a toggle-option in your xml-data: cooling_enabled. Maybe that is the connected xml-keyt, I wonder?

**Comment by robertklep** [comment]
It's an Elga firmware setting (setable through OpenTherm or with the original Honeywell Thermostat, the Anna cannot change that setting).

If you want I can try and enable cooling on the Elga, see if that changes anything?

**Comment by robertklep** [comment]
👍🏻

If you need my help with anything, let me know! FWIW, I only have an Anna and a Smile-T, not an Adam.

**Comment by bouwew** [comment]
Note to self: this Anna has no active schedule. Indicated by `active=false` and a `location_id` under `contexts/context/zone`.

**Comment by robertklep** [comment]
Your remark got me triggered, and I found the issue: for some reason, Anna had an empty schedule (I can't remember ever creating it, all my scheduling is done from HA). After remove it, no more crashes 😊

![IMG_9654](https://user-images.githubusercontent.com/1190916/199987602-22d62aa2-4e2e-469d-ab7c-f79d62ee04f5.PNG)
![IMG_9655](https://user-images.githubusercontent.com/1190916/199987610-8201e5aa-9f44-4964-9184-ddc311cdbf74.PNG)

**Comment by bouwew** [comment]
Ok, that's good to read :)

But it's good to know that an empty schedule can exits on an Anna, we should be able to detect that.

My next question is, how does your HA Anna climate-card look like? Do you see 2 set-points in the circle? One at 4/7°C and one at the heating-setpoint?

**Comment by robertklep** [comment]
Two set-points, but not quite those values:

<img width="482" alt="image" src="https://user-images.githubusercontent.com/1190916/200005674-495d66f2-a244-4c7b-bfc9-6a687bdd510f.png">

**Comment by bouwew** [comment]
Ah yes, the cooling setpoint is higher, so at 30°C.

So that looks as it should be.

Again, I still want to find a way to determine that there is no cooling function present.
I might have more question later...

**Comment by bouwew** [comment]
Good morning, I'm looking at the xml-data again, I'm seeing this:
```
			<toggle_functionality id='7de76401da0249e1b9f13759d8938749'>
				<cooling_toggle id='b556a934246f4d99979288841e7ee58a'/>
				<updated_date>2022-11-02T03:57:23.207+01:00</updated_date>
				<type>cooling_enabled</type>
				<state>off</state>
```
That points to a switch in the Plugwise App, do you see a switch in the App that might be connected to this bit of xml?

**Comment by robertklep** [comment]
The only cooling-related settings are in the Elga configuration part of the app, see below for screenshots.

None seem to be directly related to your XML excerpt (I initially thought it might have been related to the A6 dip switch setting, but the XML contains information specifically related to that).

![image](https://user-images.githubusercontent.com/1190916/200106966-82470578-8195-4e74-adf8-99727b3f337b.png)

![image](https://user-images.githubusercontent.com/1190916/200106976-80263ee5-85ea-474f-9285-5911414fec24.png)

![image](https://user-images.githubusercontent.com/1190916/200107311-3e762492-1a41-4b9b-b142-f8f41fa577aa.png)

**Comment by bouwew** [comment]
Ok thanks. I will ask Plugwise what this all means.
I was thinking, if the `cooling_enabled` key shows the state of the Elga, I could use the state of False as an indicator that cooling is not present. But since there's also a switch, the present code thinks there is an actual cooling-switch in the system.

You should see that switch in the OpenTherm device? It should be in the off-state.
Can you toggle it to on? If yes, do you see a switch change status in the Elga part of the App?

**Comment by robertklep** [comment]
> You should see that switch in the OpenTherm device? It should be in the off-state.
Can you toggle it to on? If yes, do you see a switch change status in the Elga part of the App?

At the moment only the Smile-T is connected to the Elga so can't easily check, but I'll hook up the original Honeywell thermostat and see what that gives me.

**Comment by bouwew** [comment]
That's not what I mean :)
Look at the OpenTherm device in the HA Plugwise Integration:
![image](https://user-images.githubusercontent.com/11290930/200108125-4c780d81-2b56-44b0-b819-1dab6ab42cb5.png)

**Comment by bouwew** [comment]
![image](https://user-images.githubusercontent.com/11290930/200108141-b412a926-0964-4c57-9727-e881c9b642df.png)

Do you see a cooling-switch in your system?

**Comment by bouwew** [comment]
WTH! I have a cooling-switch in my system!?!?
There's something wrong in the code ... :)

**Comment by bouwew** [comment]
Yes, it's in the xml-data of my Adam. This must recently have being added to the firmware.
I did not have it a few days ago, when the firmware was 3.6.7, now I'm on 3.7.1 beta-firmware.
Hmm, this complicates things significantly...

**Comment by robertklep** [comment]
<img width="601" alt="image" src="https://user-images.githubusercontent.com/1190916/200108459-4112b496-f136-45ad-b828-10b98ba9a9a5.png">

**Comment by bouwew** [comment]
Ok, clear thanks.

**Comment by bouwew** [comment]
Can you test plugwise-beta v0.32.0a0? This test-version also contains a fix for the Elga: the cooling-switch should be removed and a cooling_enabled binary_sensor should be shown.

**Comment by robertklep** [comment]
The switch is still there:

<img width="604" alt="image" src="https://user-images.githubusercontent.com/1190916/200381279-ab13ce35-6437-4396-b0e3-496506260ad2.png">

Integration manifest:
```
{
  "integration_manifest": {
    "domain": "plugwise",
    "name": "Plugwise Beta",
    "version": "0.32.0a0",
    "documentation": "https://github.com/plugwise/plugwise-beta",
    "after_dependencies": [
      "usb",
      "zeroconf"
    ],
    "requirements": [
      "https://test-files.pythonhosted.org/packages/ec/95/077c566aa9fe9bedc4729c5ab750bfa2a2dee6326b6009c2fd7936152274/plugwise-0.26.0a1.tar.gz#plugwise==0.26.0a0"
    ],
    "codeowners": [
      "@CoMPaTech",
      "@bouwew",
      "@brefra"
    ],
    "iot_class": "local_polling",
    "config_flow": true,
    "loggers": [
      "plugwise"
    ],
    "is_built_in": false
  },
  "data": {
    "gateway": {
      "smile_name": "Smile Anna",
      "gateway_id": "XXX",
      "heater_id": "XXX",
      "cooling_present": true,
      "notifications": {}
    },
    "devices": {
      "XXX": {
        "dev_class": "thermostat",
        "firmware": "2018-02-08T11:15:53+01:00",
        "hardware": "6539-1301-500",
        "location": "XXX",
        "model": "ThermoTouch",
        "name": "Anna",
        "vendor": "Plugwise",
        "thermostat": {
          "setpoint_low": 18.0,
          "setpoint_high": 30.0,
          "lower_bound": 4.0,
          "upper_bound": 30.0,
          "resolution": 0.1
        },
        "available": true,
        "preset_modes": [
          "vacation",
          "no_frost",
          "asleep",
          "away",
          "home"
        ],
        "active_preset": "home",
        "available_schedules": [
          "None"
        ],
        "selected_schedule": "None",
        "last_used": null,
        "mode": "heat_cool",
        "sensors": {
          "temperature": 19.1,
          "illuminance": 2.25,
          "cooling_activation_outdoor_temperature": 40.0,
          "cooling_deactivation_threshold": 4.0,
          "setpoint_low": 18.0,
          "setpoint_high": 30.0
        }
      },
      "XXX": {
        "dev_class": "gateway",
        "firmware": "4.3.12",
        "hardware": "AME Smile 2.0 board",
        "location": "XXX",
        "mac_address": "XXX",
        "model": "Gateway",
        "name": "Smile Anna",
        "vendor": "Plugwise",
        "binary_sensors": {
          "plugwise_notification": false
        },
        "sensors": {
          "outdoor_temperature": 13.6
        }
      },
      "XXX": {
        "dev_class": "heater_central",
        "location": "XXX",
        "model": "Generic heater/cooler",
        "name": "OpenTherm",
        "vendor": "Techneco",
        "maximum_boiler_temperature": {
          "setpoint": 60.0,
          "lower_bound": 0.0,
          "upper_bound": 100.0,
          "resolution": 1.0
        },
        "available": true,
        "binary_sensors": {
          "dhw_state": false,
          "heating_state": false,
          "compressor_state": false,
          "cooling_state": false,
          "slave_boiler_state": false,
          "flame_state": false
        },
        "sensors": {
          "water_temperature": 19.1,
          "intended_boiler_temperature": 0.0,
          "modulation_level": 0.0,
          "return_temperature": 18.6,
          "water_pressure": 0.5,
          "outdoor_air_temperature": 12.0
        },
        "switches": {
          "cooling_ena_switch": false,
          "dhw_cm_switch": true
        }
      }
    }
  }
}
```

**Comment by bouwew** [comment]
Damn, yes, I see why. Fixing too many bugs and implementing too many ideas at the same time.
I will let you know when I have a working release available. Might be a few days to a week...

**Comment by bouwew** [comment]
On second thought, I can quickly create a release, please test plugwise-beta v0.31.3a0.

**Comment by robertklep** [comment]
Yup, switch is unavailable now 👍🏻

<img width="595" alt="image" src="https://user-images.githubusercontent.com/1190916/200396141-f9115b54-3a21-492d-8837-5dde31234f29.png">

**Comment by bouwew** [comment]
Ok good!
I think I can still make some more improvements, I'll let you know when there's a new version to test :)

**Comment by bouwew** [comment]
Please test v0.31.3a2, the `cooling` binary_sensor should be removed. Also, a `cooling_enabled` binary_sensors has been added, showing the cooling_enabled status of the Elga.

And, you should see a single (heating-)setpoint in the climate-card.

**Comment by studiostevus** [comment]
Same issue here (same log report) as topic start. Currently v0.31.1
- dipswitch A6 on and off tested
- schedules on Anna (to test)

v0.31.3a2 not in HACS yet?

**Comment by verciIII** [comment]
Hello, I encountered the same bugs as mentioned by robertklep, and when I deleted the old shedule from last winter (not an empty one), all the errors dissapeared. I wanted to let you know this.

**Comment by studiostevus** [comment]
v0.31.3a2 seems to work!

**Comment by bouwew** [comment]
@studiostevus dipswitch A6 is for turning on cooling, next to heating?
If yes, please note, due to some HA Core restrictions the Plugwise(-beta) code cannot act on the dipswitch change.
So every time you change the dipswitch setting, make sure to reload the Plugwise(-beta) integration after the change.

With cooling-enabled, the HA climate-card will show two setpoints, with cooling not enabled the climate-card will shown one setpoint.

**Comment by bouwew** [comment]
@studiostevus @verciIII what happens when you create a new schedule? Does it still work? Or is the problem back?

**Comment by robertklep** [comment]
A6 can be used to tell Elga that the connected thermostat doesn't support cooling.

**Comment by bouwew** [comment]
Ok thanks.
If A6 is set to cooling is supported, this enables the cooling-function of the Elga, correct?

**Comment by robertklep** [comment]
@bouwew it doesn't necessarily enable or disable cooling. The manual states the following:
> Hierbij wordt gekozen of de thermostaat wel of niet automatisch kan koelen. Als de thermostaat koelen niet ondersteunt en A6 staat op ON, dan zal de Elga geen warmtevraag zien als koelvraag en koelt de Elga zolang de koeling op de voorzijde met knop 2 is vrijgegeven. De meegeleverde Honeywell Touch Modulation ondersteunt koelen (zie hoofdstuk 9.2), A6 staat dus standaard op OFF.

So basically, it determines if Elga should be "smart" about whether it should cool or not, instead of letting the thermostat determine if it should.

**Comment by studiostevus** [comment]
> @studiostevus @verciIII what happens when you create a new schedule? Does it still work? Or is the problem back?

I've created a schedule and reloaded the integration. Seems to work so far...

Another issue I'm observing is that I'm sensors are 'unavailable' periodically instead of 'off'. Not sure if this is supposed to happen...

![image](https://user-images.githubusercontent.com/104694735/200798882-3156f1bb-424d-4efd-9a67-8a8c1ac9a08e.png)

**Comment by bouwew** [comment]
@studiostevus Yes, that's a known problem. I will fix it during the weekend.

**Comment by bouwew** [comment]
#So basically, it determines if Elga should be "smart" about whether it should cool or not, instead of letting the thermostat determine if it should.

Ok, there are two ways to control how the Elga cools?

**Comment by bouwew** [comment]
>So basically, it determines if Elga should be "smart" about whether it should cool or not, instead of letting the thermostat determine if it should.
> 
 
Ok, there are two ways to control how the Elga cools?

**Comment by robertklep** [comment]
@bouwew if I understand the different manuals correctly, my Elga (I have one of the old models, I don't know if the same applies to the current models) will cool iff 1) it gets told to cool by the thermostat and 2) it's allowed to cool by pressing a button on the front of the device ("koeling vrijgeven").

If the thermostat doesn't support cooling, you can tell Elga so (with the A6 dipswitch) and _as long as it's allowed to cool_ it will try and be smart about it.

The front panel button is the "switch" that you can use to enable/disable cooling (some thermostats may also allow you to disable cooling, but that's just a thermostat-internal setting that will prevent it from sending a "cool" signal to Elga; it doesn't change anything on the Elga itself, at least AFAIK).

**Comment by bouwew** [comment]
Ok clear, thanks.

But, then I wonder if my solution in the code works properly?
@robertklep you don't use the cooling-mode at all correct?
@studiostevus @verciIII do you use the cooling-mode? And which version of the Elga do you have?

**Comment by studiostevus** [comment]
I've had the dipswitch A6 in both positions, but I don't use and don't intend to use cooling.

**Comment by studiostevus** [comment]
> @studiostevus Yes, that's a known problem. I will fix it during the weekend.

Ok looking forward to it… it also interferes with running any automations (as the thermostat may be offline when the automation triggers a change in set point)

**Comment by bouwew** [comment]
@studiostevus Plugwise-beta v0.31.1a3 is available for testing.
In this version the unavailability issue should be fixed.

**Comment by robertklep** [comment]
@bouwew shouldn't that be v0.31.3a3?

**Comment by bouwew** [comment]
Yes, v0.31.3a3 sorry :)

**Comment by bouwew** [comment]
Changelog:
- Anna + Elga: hide cooling_enable switch, (hardware-)switch is on Elga, not in Plugwise App
- Adam: improve collecting regulation_mode-related data. Fix for https://github.com/plugwise/python-plugwise/issues/240
- Anna: remove device availability, fix for https://github.com/home-assistant/core/issues/81716
- Anna + OnOff device: fix incorrect heating-state, fix for https://github.com/home-assistant/core/issues/81839
- Improve xml-data missing, raise exception with warning. Solution for https://github.com/home-assistant/core/issues/81672

**Comment by verciIII** [comment]
> @studiostevus @verciIII what happens when you create a new schedule? Does it still work? Or is the problem back?

Sorry for the late reply.

just made a generic schedule, and everything seems to work so far.

In reply to your other question, i'm using the Smile + Anna combination that was issued by Eni before Eneco bought Eni. Not Elga.

**Comment by bouwew** [comment]
Closing this, as fixed in v0.31.3
